### ISSUE TITLE
REQUEST - Notify client on authentication failure

### ISSUE THREAD

**Issue by oznogon** [daid/EmptyEpsilon#244]
When a client attempts to connect to a server of a different version, the connection fails and the server logs a message like this, as expected:

```
[INFO]  New connection: 1 waiting for authentication
[ERROR] 0:Client version mismatch: 20160519 != 20160518
```

However, there are no visible or logged messages on the client:

```
[INFO]  Reading servers from master server
[INFO]  Loaded: gui/TextEntryBackground.focused with 0 sprites
[INFO]  GameClient: Connected, waiting for authentication
[INFO]  Reading servers from master server
(END)
```

Since the server version isn't always known to the client, client connection issues can be difficult to resolve.

**Comment by daid** [comment]
Odd, I though I made the code so that mismatched versions are not displayed in server browsing...

**Comment by oznogon** [comment]
They don't show up, but a user can still attempt to connect by IP. If the server's password-protected, a user on a mismatched version can still get the server's password prompt and authenticate, but the connection fails silently on their end.

This affects LAN play too, FWIW. Autodiscovery is working as expected, there's just no client-side logging if clients aren't relying on autodiscovery to connect.

**Comment by oznogon** [comment]
Necromancing this one; the connection happens in SeriousProton's `multiplayer_client`: https://github.com/daid/SeriousProton/blob/eff2a15/src/multiplayer_client.cpp#L58-L185

If I'm reading it right:

1. Client requests auth by sending a packet to the server
2. If auth is required, the server's packet comes back with the server version, and potentially a password requirement
3. If there's a password requirement, wait for the client to provide a password
    1. If the password is provided, send it
4. If there's no password requirement, or if a password request is authenticated, proceed as if authenticated and send the client version number

The solution here at least seems to be at step 2:

```
packet >> server_version >> require_password;
```

Compare `server_version` to `version_number` (the client version) at this point and at least log this on the client side. IIRC also ignore if the server version is 0 for dev/test purposes:

```
packet >> server_version >> require_password;

if (server_version != 0 && server_version != version_number)
{
  LOG(INFO) << "Server version " << server_version << " does not match client version " << version_number;
}
```

This at least gives the client a hint as to why the server might then end the connection. More severely, the client could also break here and the logged message would be an ERROR instead.

**Comment by oznogon** [comment]
https://github.com/daid/SeriousProton/pull/39 logs the mismatch on the client per the above guess. Worked when testing with a 20191101 client connecting to a master branch build.

A GUI notification seems doable, but a bit tougher and EE/client specific since the server doesn't respond that auth failed, it just sends another auth request.

**Comment by JonathanILevi** [comment]
@daid this is solved by daid/SeriousProton#39.  (although not tested myself)

See #707 for an issue for adding a GUI message.

**Comment by daid** [comment]
Oh, I reopened this because I thought it was about a UI notification. But the other one is.
